test/integration/card_details_send_epdq_3ds2_data_ft_tests.js

```diff
@@ -0,0 +1,189 @@
+  addressCountry: addressCountry,
+  jsNavigatorLanguage: jsNavigatorLanguage,
+  jsScreenColorDepth: jsScreenColorDepth,
+  jsScreenHeight: jsScreenHeight,
+  jsScreenWidth: jsScreenWidth,
+  jsTimezoneOffsetMins: jsTimezoneOffsetMins
+}
+
+const expectedDataSentToConnector = {
+  card_number: cardNumber,
+  cvc: cvc,
+  card_brand: 'visa',
+  expiry_date: expiryMonth + '/' + expiryYear,
+  cardholder_name: cardholderName,
+  card_type: 'DEBIT',
+  corporate_card: false,
+  prepaid: 'NOT_PREPAID',
+  address: {
+    line1: addressLine1,
+    line2: addressLine2,
+    city: addressCity,
+    postcode: addressPostcode,
+    country: addressCountry
+  },
+  js_navigator_language: jsNavigatorLanguage,
+  js_screen_color_depth: jsScreenColorDepth,
+  js_screen_height: jsScreenHeight,
+  js_screen_width: jsScreenWidth,
+  js_timezone_offset_mins: jsTimezoneOffsetMins,
+  ip_address: '127.0.0.1'
+}
+
+function connectorExpects (data) {
+  return mockServer.post(connectorChargePath + chargeId + '/cards', body => {
+    // Strip non body data (header data) before comparing
+    delete body.accept_header
+    delete body.user_agent_header
+    console.log('COMPARING: ')
+    console.log(JSON.stringify(body))
+    console.log('WITH: ')
+    console.log(JSON.stringify(data))
+    console.log('RETURNING: ' + _.isEqual(body, data))
+    return _.isEqual(body, data)
+  })
+}
+
+describe('Enter card details page — sending ePDQ 3DS2 additional data to connector', function () {
+  beforeEach(function () {
+    nock.cleanAll()
+    mockServer = nock(process.env.CONNECTOR_HOST)
+  })
+
+  before(function () {
+    // Disable logging.
+    logger.level = 'none'
+  })
+
+  it('should send card data including all ePDQ 3DS2 additional data', function (done) {
+    const cookieValue = cookie.create(chargeId)
+
+    nock(process.env.CONNECTOR_HOST)
+      .patch('/v1/frontend/charges/' + chargeId)
+      .reply(200)
+    defaultConnectorResponseForGetCharge(chargeId, State.ENTERING_CARD_DETAILS, gatewayAccountId)
+    defaultAdminusersResponseForGetService(gatewayAccountId)
+    mockSuccessCardIdResponse()
+
+    const dataSentToConnector = expectedDataSentToConnector
+
+    connectorExpects(dataSentToConnector).reply(200)
+
+    const formData = formSubmissionData
+
+    postChargeRequest(app, cookieValue, formData, chargeId)
+      .expect(303)
+      .expect('Location', frontendCardDetailsPath + '/' + chargeId + '/confirm')
+      .end(done)
+  })
+
+  it('should send card data including ePDQ 3DS2 additional data without jsScreenColorDepth', function (done) {
+    const cookieValue = cookie.create(chargeId)
+
+    nock(process.env.CONNECTOR_HOST)
+      .patch('/v1/frontend/charges/' + chargeId)
+      .reply(200)
+    defaultConnectorResponseForGetCharge(chargeId, State.ENTERING_CARD_DETAILS, gatewayAccountId)
+    defaultAdminusersResponseForGetService(gatewayAccountId)
+    mockSuccessCardIdResponse()
+
+    const dataSentToConnector = expectedDataSentToConnector
```

Comment:
I think this assignment doesn't need to happen either. I did a simple test:
```
> var list1 = [1,2,3]
undefined
> var list2 = list1
undefined
> list2[1] = 5
5
> list1
[ 1, 5, 3 ]
> list2
[ 1, 5, 3 ]
```
Reassigning it doesn't create a new object assigned to a new variable. The original object here `expectedDataSentToConnector` will change when `dataSentToConnector` is changed.

Reply:
Yeah, the intention wasn’t to clone the object.

Honestly, some of the code in this test was ~copied~ inspired by other tests and in the other tests the equivalent of `expectedDataSentToConnector` was a function rather than a variable (it took some arguments and returned an appropriate object), so the whole assignment thing is just an artefact of that. Will remove.